Allow font-variant east-asian shorthand in any position

mdubet opened the pull request:
#### fd7b349b71efa48b759a7ceddc8ddc336874c895
<pre>
Allow font-variant east-asian shorthand in any position
<a href="https://bugs.webkit.org/show_bug.cgi?id=245972">https://bugs.webkit.org/show_bug.cgi?id=245972</a>

Reviewed by Darin Adler.

The property font-variant-east-asian can be used in any position
in the font-variant shorthand ; it should not check
that the range is finished.

<a href="https://drafts.csswg.org/css-fonts-4/#font-variant-prop">https://drafts.csswg.org/css-fonts-4/#font-variant-prop</a>

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:

Add tests which fail before this change.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid.html:

Imported from WPT.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-valid-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantEastAsian):

Canonical link: <a href="https://commits.webkit.org/255134@main">https://commits.webkit.org/255134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/212360014c00f59f3d14d150aed7c9fa552a0797

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/568 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101194 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161210 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/586 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97555 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97126 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27352 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82316 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35601 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33387 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/17098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3569 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37187 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39107 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/36215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -702,70 +702,77 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
+        bool hasParsedSomething = false;
 
-        auto id = range.peek().id();
+        for (;;) {
```
```suggestion
        while (true) {
```

nit: `while (true)` is much more largely used in WebKit: https://searchfox.org/wubkat/search?q=while+%28true%29&path=.cpp&case=false&regexp=false

compared to `for (;;)`: https://searchfox.org/wubkat/search?q=for+%28%3B%3B%29+%7B&path=cpp&case=false&regexp=false

mdubet replied:
Changed to `while (true)`. The repartition of usage in our codebase (1/3 vs 2/3) is not great, we should agree on one of them and add a style checker rule for it.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -702,70 +702,77 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
+                return hasParsedSomething;
+
+            auto id = range.peek().id();
+            switch (id) {
```
```suggestion
            switch (range.peek().id()) {
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -702,70 +702,77 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
+                    return false;
+                ruby = FontVariantEastAsianRuby::Yes;
+                break;
+            default:
+                return hasParsedSomething;
```
My worry is that your logic may allow `font-variant-east-asian: full-width garbage` (though I haven't tested)

The test for invalid values doesn't cover that case (only covers duplicated variant/ruby/width or normal followed by smtg incorrect): https://searchfox.org/wubkat/source/LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid.html

Might be nice to test `font-variant-east-asian: full-width garbage` (where garbage is neither variant/ruby/width), or something like  `font-variant-east-asian: full-width none`, and add these to the WPT (I can help with exporting).

mdubet replied:
It works. The range is always checked by the caller function (`parseSingleValue` or `parseShorthandXXX`) which has more context about "should it be the end of a range or not?".

I've added tests.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -702,70 +702,77 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
-    while (!range.atEnd()) {
-        if (range.peek().type() != IdentToken)
-            return nullptr;
+    auto parseSomethingWithoutError = [&range, &variant, &width, &ruby] () {
+        bool hasParsedSomething = false;
```
Mostly because I'm not a huge fan of `parseSomethingWithoutError` as name, I would suggest doing something like `while (!shouldFail)` and set `shouldFail` appropriately, then add early `return nullptr` if `shouldFail` is true.

mdubet replied:
I think it would make the workflow more complicated (and kinda defeat the purpose of the lambda).

nt1m reviewed (commented):
I'm wondering if `font-variant-east-asian: full-width garbage` or `font-variant-east-asian: full-width none` is correctly handled here. We've had a similar bug with the `list-style` shorthand, where it supported `list-style: none garbage` because it didn't do a proper range end check.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -702,70 +702,76 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
 
-        auto id = range.peek().id();
+        while (true) {
+            if (range.peek().type() != IdentToken)
+                return hasParsedSomething;
```
I'm now wondering how it handles `full-width 30px` or something similar (where the first part sets `hasParsedSomething` to `true` and the second part isn't an ident).

mdubet replied:
It works also, cf my previous comment (it is the caller function which handles the rest of the range).

nt1m commented on LayoutTests/fast/text/font-variant-shorthand.html:
```diff
@@ -156,6 +172,10 @@
 shouldBeEqualToString("window.getComputedStyle(document.getElementById('t19')).getPropertyValue('font-variant')", "normal");
 shouldBeEqualToString("window.getComputedStyle(document.getElementById('t19')).getPropertyValue('font-variant-caps')", "normal");
+shouldBeEqualToString("window.getComputedStyle(document.getElementById('t20')).getPropertyValue('font-variant')", "common-ligatures simplified");
+shouldBeEqualToString("window.getComputedStyle(document.getElementById('t21')).getPropertyValue('font-variant-east-asian')", "normal");
+shouldBeEqualToString("window.getComputedStyle(document.getElementById('t22')).getPropertyValue('font-variant-east-asian')", "normal");
+shouldBeEqualToString("window.getComputedStyle(document.getElementById('t23')).getPropertyValue('font-variant-east-asian')", "normal");
```
This doesn't really test if parsing is invalid, especially since getComputedStyle() will normalize values, etc.

I've sent a WPT PR: https://github.com/web-platform-tests/wpt/pull/36254/files , feel free to import it to the WebKit copy once it's merged.

This is what the test file uses: https://searchfox.org/wubkat/rev/e8e9be7353991aee7df45aa62f0e0148f16e6f9f/LayoutTests/imported/w3c/web-platform-tests/css/support/parsing-testcommon.js#29-38

mdubet replied:
Imported 👍

mdubet committed:
Allow font-variant east-asian shorthand in any position
https://bugs.webkit.org/show_bug.cgi?id=245972

Reviewed by Darin Adler.

The property font-variant-east-asian can be used in any position
in the font-variant shorthand ; it should not check
that the range is finished.

https://drafts.csswg.org/css-fonts-4/#font-variant-prop

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:

Add tests which fail before this change.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid.html:

Imported from WPT.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-valid-expected.txt:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantEastAsian):

Canonical link: https://commits.webkit.org/255134@main
